docs/AQL/grammar/AqlParser.g4

```diff
@@ -0,0 +1,214 @@
+    | whereExpr OR whereExpr
+    | OPEN_PAR whereExpr CLOSE_PAR
+    ;
+
+orderByExpr
+    : identifiedPath order=(DESCENDING|DESC|ASCENDING|ASC)?
+    ;
+
+
+columnExpr
+    : identifiedPath
+    | aggregateFunctionCall
+    | functionCall
+    ;
+
+containsExpr
+    : classExprOperand (CONTAINS containsExpr)?
+    | containsExpr AND containsExpr
+    | containsExpr OR containsExpr
+    | OPEN_PAR containsExpr CLOSE_PAR
+    ;
+
+identifiedExpr
+    : EXISTS identifiedPath
+    | identifiedPath COMPARISON_OPERATOR identifiedOperand
+    | functionCall COMPARISON_OPERATOR identifiedOperand
+    | identifiedPath LIKE likeOperand
+    | identifiedPath MATCHES OPEN_CURLY matchesOperand CLOSE_CURLY
+    ;
+
+identifiedOperand
+    : primitive
+    | PARAMETER
+    | identifiedPath
+    | functionCall
+    ;
+
+identifiedPath
+    : IDENTIFIER pathPredicate? (SLASH objectPath)?
+    ;
+
+pathPredicate
+    : OPEN_BRACKET (standardPredicate | archetypePredicate | nodePredicate) CLOSE_BRACKET
+    ;
+
+standardPredicate
+    : objectPath COMPARISON_OPERATOR pathPredicateOperand
+    ;
+
+archetypePredicate
+    : ARCHETYPEID
+    | PARAMETER
+    ;
+
+nodePredicate
+    : NODEID (COMMA (STRING|PARAMETER))?
+    | ARCHETYPEID (COMMA (STRING|PARAMETER))?
+    | PARAMETER
+    | objectPath COMPARISON_OPERATOR pathPredicateOperand
+    | objectPath MATCHES REGEXPATTERN
+    | nodePredicate AND nodePredicate
+    | nodePredicate OR nodePredicate
+    ;
+
+pathPredicateOperand
+    : primitive
+    | objectPath
+    | PARAMETER
+    ;
+
+
+objectPath
+    : pathPart (SLASH pathPart)*
+    ;
+pathPart
+    : IDENTIFIER pathPredicate?
+    ;
+
+likeOperand
+    : STRING
+    | PARAMETER
+    ;
+matchesOperand
+    : valueListItem (COMMA valueListItem)*
+    | terminologyFunction
+    | URIVALUE
+    ;
+valueListItem
+    : primitive
+    | PARAMETER
+    | terminologyFunction
+    ;
+
+primitive
+    : STRING
+    | INTEGER
+    | FLOAT
+    | DOUBLE
+    | DATE
+    | BOOLEAN
+    ;
+
+
+
+classExprOperand
+    : IDENTIFIER IDENTIFIER? // RM_TYPE_NAME variable
+    | archetypedClassExpr
+    | versionedClassExpr
+    | versionClassExpr
+    ;
+
+// RM_TYPE_NAME [archetype_id]
+// RM_TYPE_NAME variable [archetype_id]
+archetypedClassExpr
+    : IDENTIFIER IDENTIFIER? archetypePredicate
+    ;
+versionedClassExpr
+    : VERSIONED_OBJECT IDENTIFIER? (OPEN_BRACKET standardPredicate OPEN_BRACKET)?
+    ;
+versionClassExpr
+    : VERSION IDENTIFIER? (OPEN_BRACKET (standardPredicate|versionPredicate) OPEN_BRACKET)?
+    ;
+versionPredicate
+    : LATEST_VERSION
+    | ALL_VERSIONS
+    ;
+
+functionCall
+    : terminologyFunction
+    | function
+    ;
+
+function
+    : IDENTIFIER OPEN_PAR functionArg (COMMA functionArg)* CLOSE_PAR
+    ;
+
+functionArg
+    : primitive
+    | identifiedPath
+    | PARAMETER
+    | functionCall
+    ;
+
+aggregateFunctionCall
+    : COUNT OPEN_PAR (DISTINCT? identifiedPath | STAR) CLOSE_PAR    #countFunction
```

Comment:
I would have expected this to be taken care of by one rule of the form:
```
aggregateFunctionCall:
      aggregateFunc OPEN_PAREN identifiedObject CLOSE_PAREN
    ;

aggregateFunc:
      MIN 
    | MAX
    | SUM
    | AVG
    ;

Reply:
Agree.

Reply:
... ok, but count() has a different signature - should that be kept separately?
In the future there might be other aggregate function to add here, with other signatures (when compared to min(), max(), etc)

Reply:
Ok, well, there are a few ways to go about this, as far as I know, and a lot of it comes down to 'taste'. 

My background is more in general purpose language development, and there function call signatures are usually handled very freely, like this:
```Java
functionCall
  : FUNCTION_IDENTIFIER OPEN_PAREN (argument (COMMA argument)*)? CLOSE_PAREN
  ;
```
The details of `argument` are not really relevant here, and `FUNCTION_IDENTIFIER` is just a lexer rule that defines the specific rules for function names (which might or might not be different from general identifiers). Specific function names, even in the case of build-in functions, are not defined, and all those things are handled in the semantic analysis phase (is the value of `FUNCTION_IDENTIFIER` a known function, even for built-in functions, and is the cardinality and type of arguments correct, things like that), and not in the grammar. This allows for a lot of flexibility, but also leaves a lot to the implementer of the semantic analysis.

Now as far as I can tell, SQL-like languages tend to be a bit more specific in the grammar. The style used in the grammar of mysql-workbench follows the style we currently use (at least for some constructs) as you can see here: 
https://github.com/mysql/mysql-workbench/blob/7e5826afd6198a89d6a43624101bbeeb163eb2d7/library/parsers/grammars/MySQLParser.g4#L2474
Basically they do what we do in the current code, even if the signatures are the same (aside from the function name). They probably also have an alternative somewhere for user-defined functions, as that is something mysql supports. That likely follows the pattern I described first.

Finally, you can basically minimize duplication by doing what @wolandscat suggests. It is a style also a lot of people prefer, because it may look a bit cleaner. Note that, if done correctly, there is no difference in what language the last two alternatives accept (the first one is much more flexible).

Which one to chose is a bit of a matter of taste (in particular in the latter two alternatives). The first option makes it very easy for language implementers to add functions in a sort-of consistent style without having to modify the grammar. The latter two options make it more clear in the grammar which functions are available, but is a bit more constraining.

It is also perfectly fine to mix-and-match elements from these styles, for example, use the third alternative for MIN, MAX, AVG, and SUM (as they have exactly the same signature w.r.t. parameters), mixing it with alternative 2 for COUNT, and maybe even 1 as an 'extension point'.

I'd say for documentation purposes one of alternative 2 / 3 might be clearer, as it shows in the grammar, but you'd still have to specifiy the semantics in the accompanying specification anyway. 

Once again, note that when some party implements this, they might modify the grammar anyway to better suit their style of implementation (while of course still ensuring that their specific grammar accepts the same language as the 'reference grammar' here).

Reply:
Ok, I made some changes, hopefully reducing complexity while still maintaining readability). Please review it again in the next round @teunvanhemert